bookdown/03-optimization-searchspace.Rmd

```diff
@@ -0,0 +1,223 @@
+
+```{r 03-optimization-searchspace-004}
+tune_ps = ps(
+  cost = p_dbl(-1, 1, trafo = function(x) 10^x),
+  kernel = p_fct(c("polynomial", "radial"))
+)
+rbindlist(generate_design_grid(tune_ps, 3)$transpose())
+```
+
+It is even possible to attach another transformation to the `r ref("ParamSet")` as a whole that gets executed after individual parameter's transformations were performed. It is given through the `.extra_trafo` argument and should be a function with parameters `x` and `param_set` that takes a list of parameter values in `x` and returns a modified list. This transformation can access all parameter values of an evaluation and modify them with interactions. It is even possible to add or remove parameters. (The following is a bit of a silly example.)
+
+```{r 03-optimization-searchspace-005}
+tune_ps = ps(
+  cost = p_dbl(-1, 1, trafo = function(x) 10^x),
+  kernel = p_fct(c("polynomial", "radial")),
+  .extra_trafo = function(x, param_set) {
+    if (x$kernel == "polynomial") {
+      x$cost = x$cost * 2
+    }
+    x
+  }
+)
+rbindlist(generate_design_grid(tune_ps, 3)$transpose())
+```
+
+The available types of search space parameters is limited: continuous, integer, discrete, and logical scalars. There are many machine learning algorithms, however, that take parameters of other types, for example vectors or functions. These can not be defined in a search space `r ref("ParamSet")`, and they are often given as `r ref("ParamUty")` when defining a `r ref("Learner")`'s `r ref("ParamSet")`. When trying to tune over these hyperparameters, it is necessary to perform a Transformation that changes the type of a parameter.
+
+An example is the `class.weights` parameter of the SVM, which takes a named vector of class weights with one entry for each target class. The trafo that would tune `class.weights` for the `r ref("mlr_tasks_spam", "tsk(\"spam\")")` dataset could be:
+
+```{r 03-optimization-searchspace-006}
+tune_ps = ps(
+  class.weights = p_dbl(0.1, 0.9, trafo = function(x) c(spam = x, nonspam = 1 - x))
+)
+generate_design_grid(tune_ps, 3)$transpose()
+```
+
+(We are omitting `rbindlist()` in this example because it breaks the vector valued return elements.)
+
+### Automatic Factor Level Transformation {#autolevel}
+
+A common use-case is the necessity to specify a list of values that should all be tried (or sampled from). It may be the case that a hyperparameter takes on function values and a certain list of functions should be tried. Or it may be that a choice of special numeric values should be used. For this, the `p_fct` constructor's `level` argument may be a value that is not a `character` vector, but something else. If, for example, only the values `0.1`, `3`, and `10` should be tried for the `cost` parameter, even when doing random search, then the following search space would achieve that:
```

Comment:
```suggestion
A common use-case is the necessity to specify a list of values that should all be tried (or sampled from). It may be the case that a hyperparameter accepts function objects as values and a certain list of functions should be tried. Or it may be that a choice of special numeric values should be tried. For this, the `p_fct` constructor's `level` argument may be a value that is not a `character` vector, but something else. If, for example, only the values `0.1`, `3`, and `10` should be tried for the `cost` parameter, even when doing random search, then the following search space would achieve that:
```